voxblox/src/integrator/tsdf_integrator.cc

```diff
@@ -0,0 +1,567 @@
+#include "voxblox/integrator/tsdf_integrator.h"
+
+namespace voxblox {
+
+TsdfIntegratorBase::TsdfIntegratorBase(const Config& config,
+                                       Layer<TsdfVoxel>* layer)
+    : config_(config), layer_(layer) {
+  DCHECK(layer_);
+
+  voxel_size_ = layer_->voxel_size();
+  block_size_ = layer_->block_size();
+  voxels_per_side_ = layer_->voxels_per_side();
+
+  voxel_size_inv_ = 1.0 / voxel_size_;
+  block_size_inv_ = 1.0 / block_size_;
+  voxels_per_side_inv_ = 1.0 / voxels_per_side_;
+
+  if (config_.integrator_threads == 0) {
+    LOG(WARNING) << "Automatic core count failed, defaulting to 1 threads";
+    config_.integrator_threads = 1;
+  }
+}
+
+// Thread safe.
+inline bool TsdfIntegratorBase::isPointValid(const Point& point_C,
+                                             bool* is_clearing) const {
+  DCHECK(is_clearing != nullptr);
+  const FloatingPoint ray_distance = point_C.norm();
+  if (ray_distance < config_.min_ray_length_m) {
+    return false;
+  } else if (ray_distance > config_.max_ray_length_m) {
+    if (config_.allow_clear) {
+      *is_clearing = true;
+      return true;
+    } else {
+      return false;
+    }
+  } else {
+    *is_clearing = false;
+    return true;
+  }
+}
+
+// Will return a pointer to a voxel located at global_voxel_idx in the tsdf
+// layer. Thread safe.
+// Takes in the last_block_idx and last_block to prevent unneeded map lookups.
+// If the block this voxel would be in has not been allocated, a voxel in
+// temp_voxel_storage is allocated and returned instead.
+// This can be merged into the layer later by calling
+// updateLayerWithStoredVoxels(temp_voxel_storage)
+inline TsdfVoxel* TsdfIntegratorBase::findOrTempAllocateVoxelPtr(
+    const VoxelIndex& global_voxel_idx, Block<TsdfVoxel>::Ptr* last_block,
+    BlockIndex* last_block_idx, VoxelMap* temp_voxel_storage) const {
+  DCHECK(temp_voxel_storage != nullptr);
+
+  const BlockIndex block_idx =
+      getBlockIndexFromGlobalVoxelIndex(global_voxel_idx, voxels_per_side_inv_);
+
+  if (block_idx != *last_block_idx) {
+    *last_block = layer_->getBlockPtrByIndex(block_idx);
+    *last_block_idx = block_idx;
+    if (*last_block != nullptr) {
+      (*last_block)->updated() = true;
+    }
+  }
+
+  // If no block at this location currently exists, we allocate a temporary
+  // voxel that will be merged into the map later
+  if (*last_block == nullptr) {
+    return &((*temp_voxel_storage)[global_voxel_idx]);
+  } else {
+    const VoxelIndex local_voxel_idx =
+        getLocalFromGlobalVoxelIndex(global_voxel_idx, voxels_per_side_);
+
+    return &((*last_block)->getVoxelByVoxelIndex(local_voxel_idx));
+  }
+}
+
+// NOT thread safe
```

Comment:
Sure, if it is too much overhead, let's not do it. But just FYI we have a reader-writer lock implementation in aslam_cv2: [reader-writer-lock.h](https://github.com/ethz-asl/aslam_cv2/blob/master/aslam_cv_common/include/aslam/common/reader-writer-lock.h) and [reader-first-reader-writer-lock.h](https://github.com/ethz-asl/aslam_cv2/blob/master/aslam_cv_common/include/aslam/common/reader-first-reader-writer-lock.h)